### PR TITLE
feat(mcp): Streamable HTTP and SSE external federation transports (mcp-unified 0.3.0)

### PR DESCRIPTION
**Pull request by rmusser01**
## Why

mcp-unified's federation layer could only reach upstream MCP servers it spawns as local processes (`stdio`); the entire hosted-MCP ecosystem (Linear, Sentry, Notion, and every other URL-served MCP endpoint) was unreachable. This is the server half of tldw_chatbook's TASK-25900 / ADR-111 Option C: transports live in the package the chatbook already pins, chatbook then only does wiring.

## What

**Two new package-owned transports** in `apps/mcp-unified/src/mcp_unified/federation/http_transport.py`, both implementing the `ExternalFederationTransport` contract with stdio-parity semantics (same tools/list normalization, upstream tool errors as `is_error` results, same health_check keys, secret-free reason-coded exceptions):

- `StreamableHttpExternalTransport` — MCP Streamable HTTP (2025-03-26): JSON-RPC POSTs against one endpoint, handles both `application/json` and `text/event-stream` response bodies, `mcp-session-id` capture/propagation, `MCP-Protocol-Version` echo, DELETE session teardown, and transparent reinitialize+retry on 404 session expiry.
- `SseExternalTransport` — legacy HTTP+SSE: persistent GET stream, endpoint-event resolution **pinned to the stream origin**, response correlation by request id, pending requests failed on stream loss.

**Model/registry changes:**
- `ExternalServerDefinition`: transports `streamable_http`/`sse` (url required) + static `headers` field (auth for hosted servers).
- Gateway management: `headers` patchable (token rotation), header **values redacted (`"***"`) in every list/show/create/patch response** while the store keeps real values; enabled URL-transport servers require `http(s)://` URLs.
- `create_external_transport` dispatches to the new factory; stdio/websocket behavior untouched.

**Failure honesty** (distinct reason codes for downstream readiness states): `auth_required` (401/403), `tls_failed` (`ssl.SSLError` in the cause chain), `connect_failed`, `request_timeout`, `connection_closed`, `upstream_http_error`, `invalid_endpoint`, `insecure_url` (credential headers over plain http to non-loopback are refused).

**Hardening** (from a pre-PR adversarial review round; every finding fixed and pinned red-first):
- SSE endpoint events can no longer redirect credentialed POSTs cross-origin (confirmed-exploitable pre-fix).
- SSE-framed POST responses are bounded by total request timeout (a keepalive-dripping server previously hung the call forever and wedged `ExternalFederationManager` via its lock).
- Public operations serialize behind a request lock (stdio parity), so reconnects can't destroy in-flight requests.

## Tests

- 30 tests in `test_http_external_transport.py` run both transports against **real uvicorn stub MCP servers on ephemeral ports** (httpx's ASGI test transport buffers streams, so SSE needs real sockets): session enforcement, SSE-framed responses, runtime-auth header merge, request serialization, timeout bounding, cross-origin endpoint rejection, session-expiry recovery, SSE parser edge cases, reason-code mapping.
- 3 new gateway registry tests (redaction, header patch, URL validation). Affected suites: 165 passed.
- Full `MCP_unified/tests` dir: failure set **byte-identical** to a clean checkout of dev (same 22F/8E env-baseline set; by-name diff).

## Notes for release

- Version bumped 0.2.1 → 0.3.0 (doc copies synced). tldw_chatbook will bump its `mcp` extra pin after release.
- Compatibility: definitions saved by 0.3.0 serialize the new `headers` field; a 0.2.x reader (`extra="forbid"`) rejects such rows — worth a release-note line.
- Scope: the YAML/JSON config-file registry (`ExternalServerRegistryConfig`) still declares only `stdio`/`websocket`; URL transports are storage-registry-only for now (documented in USER_GUIDE).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01Bw1uiUyYcAWcrKVHGuJCXi

---
## Summary by cubic
Adds Streamable HTTP and legacy SSE transports to mcp-unified's federation layer, so hosted MCP servers (Linear, Sentry, and other URL-served endpoints) are reachable for the first time — previously only locally spawned `stdio` servers worked. This is the server half of TASK-25900 / ADR-111 Option C.

**New Features**
- Adds `StreamableHttpExternalTransport` (single-endpoint JSON-RPC with JSON or SSE-framed responses, session capture and DELETE teardown) and `SseExternalTransport` (persistent GET stream with request-id correlation).
- `ExternalServerDefinition` gains `streamable_http`/`sse` transports and a static `headers` field; brokered per-call credentials merge their headers into tool calls.
- Failures map to distinct secret-free reason codes (`auth_required`, `tls_failed`, `connect_failed`, `request_timeout`, `connection_closed`).
- SSE endpoint events are pinned to the stream origin, SSE responses are bounded by the request timeout, and public operations serialize behind a request lock that `close()` also honors.
- Management responses redact `headers` values (`"***"`); the FastAPI patch body supports the new transports and `headers`, and header changes restart the active transport so tokens rotate cleanly.
- Credential headers over plain `http://` to non-loopback hosts are refused, and streamable JSON responses must match the request id.

**Migration**
- Version bumps to 0.3.0; definitions saved by 0.3.0 include the `headers` field, which a 0.2.x reader rejects.
- The YAML/JSON config-file registry still only declares `stdio`/`websocket`; URL transports are storage-registry-only.

<sup>Written for commit 27cd02755563d326d218787ac60e0bc8e0524cb8. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/rmusser01/tldw_server/pull/2861?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 1]
```diff
@@ -0,0 +1,1049 @@
...
+_PROTOCOL_VERSION_HEADER = "mcp-protocol-version"
+_ACCEPT_BOTH = "application/json, text/event-stream"
+_SENSITIVE_HEADER_NAMES = frozenset(
+    {"authorization", "proxy-authorization", "cookie", "x-api-key", "api-key"}
+)
+_LOOPBACK_HOSTS = frozenset({"localhost", "127.0.0.1", "::1"})
+
+
+class HttpExternalTransportError(RuntimeError):
+    """Raised when an HTTP transport operation fails without exposing secrets."""
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py:46]: Declined with evidence: the package has no central exception module — per-module reason-coded errors are its existing convention (`StdioExternalTransportError` in stdio_transport.py, `FederationPolicyDenied` in federation/models.py, storage errors in interfaces/storage.py). `HttpExternalTransportError` follows the stdio transport's pattern exactly and is exported from `mcp_unified.federation` like its siblings.

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 2]
```diff
@@ -0,0 +1,1049 @@
...
+        self._health_timeout_s = self._positive_timeout(health_timeout_s, "health_timeout_s")
+        self._connect_lock = asyncio.Lock()
+        self._request_lock = asyncio.Lock()
+        self._next_request_id = 1
+        self._initialized = False
+
+    def _take_request_id(self) -> int:
+        request_id = self._next_request_id
+        self._next_request_id += 1
+        return request_id
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py:211]: Fixed in 27cd027555 — every helper in http_transport.py now carries a docstring (26 added), including `_take_request_id`, `_error`, `_request`, `_pump_events`, and `_teardown`.

File: tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+)
+from mcp_unified.federation.models import BrokeredExternalCredential
+from mcp_unified.federation.stdio_transport import create_external_transport
+from mcp_unified.storage import ExternalServerDefinition
+
+_SESSION_ID = "stub-session-123"
+
+
+def _tool_rows() -> list[dict[str, Any]]:
+    return [
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py:32]: Fixed in 27cd027555 — docstrings added to all test helpers, stub methods, test classes, and test functions in the module (41 added).

File: tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py [hunk 4]
```diff
@@ -0,0 +1,707 @@
...
+            "result": {"contents": [{"uri": params.get("uri"), "text": "hello"}]},
+        }
+    return {
+        "jsonrpc": "2.0",
+        "id": request_id,
+        "error": {"code": -32601, "message": f"unknown method {method}"},
+    }
+
+
+async def _read_body(receive) -> bytes:
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py]: Fixed in 27cd027555 — ASGI callables are annotated via `_Scope`/`_Receive`/`_Send` aliases (`Callable[[], Awaitable[dict[str, Any]]]` etc.) on `__call__`, `_read_body`, `_send_json`, `_wait_disconnect`, and `_run_stub`.

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 5]
```diff
@@ -0,0 +1,1049 @@
...
+                    payload = json.loads(data)
+                except json.JSONDecodeError:
+                    continue
+                if not isinstance(payload, dict):
+                    continue
+                future = self._pending.pop(payload.get("id"), None)
+                if future is not None and not future.done():
+                    future.set_result(payload)
+        except Exception:  # noqa: BLE001 - stream teardown races vary by backend.
+            pass
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py]: Fixed in 27cd027555 — the reader now logs the failure via loguru (`server_id` + exception type only, no payloads or headers, keeping the secret-free error posture) before failing pending requests with `connection_closed`.

File: tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py [hunk 6]
```diff
@@ -0,0 +1,707 @@
...
+
+        stub = StreamableHttpStub(keepalive_forever=True)
+        async with _run_stub(stub) as url:
+            transport = StreamableHttpExternalTransport(
+                _http_server_definition(url), request_timeout_s=1.0
+            )
+            started = monotonic()
+            try:
+                with pytest.raises(HttpExternalTransportError) as excinfo:
+                    await transport.list_tools()
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply tldw_Server_API/app/core/MCP_unified/tests/test_http_external_transport.py:575]: Addressed in 27cd027555 — the wall-clock ceiling is relaxed to 20s against a 1s configured timeout (20x headroom; the assertion pins boundedness/reason code, not scheduler timing). Kept rather than removed because without it a regression to an unbounded read would hang the test instead of failing it.

File: apps/mcp-unified/src/mcp_unified/gateway/external_registry.py [hunk 7]
```diff
@@ -31,6 +31,7 @@
         "cwd",
         "env_allowlist",
         "credential_slots",
+        "headers",
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/gateway/external_registry.py:34]: Fixed in 27cd027555 — `headers` is included in `_runtime_signature` (sorted items), so `_definition_changed` flags a rotation and reconcile restarts the transport with the fresh token. Pinned by `test_runtime_signature_detects_header_changes`.

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 8]
```diff
@@ -0,0 +1,1049 @@
...
+        }
+        if extra_headers:
+            headers.update(extra_headers)
+        encoded = self._encode_payload(payload, method)
+        try:
+            response = await client.post(
+                endpoint,
+                content=encoded,
+                headers=headers,
+                timeout=self._request_timeout_s,
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py:1052]: Fixed in 27cd027555 — `SseExternalTransport._request` now wraps the POST phase and the correlated-response wait in a single `asyncio.wait_for` (`_round_trip`), so the configured deadline covers the whole operation including a keepalive-dripping POST response. Pinned red-first by `test_dripping_post_response_is_bounded_by_request_timeout` (the RED run hung until pytest-timeout killed it).

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 9]
```diff
@@ -0,0 +1,1049 @@
...
+            if response.status_code >= 400:
+                raise self._status_error(response.status_code, method)
+            session_id = response.headers.get(_SESSION_HEADER)
+            if session_id:
+                self._session_id = session_id
+            content_type = response.headers.get("content-type", "")
+            if content_type.startswith("text/event-stream"):
+                return await self._read_sse_response(response, request_id, method)
+            body = await response.aread()
+            return self._decode_json_response(body, method)
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py]: Fixed in 27cd027555 — plain JSON responses must now carry the matching JSON-RPC id or the request fails with `invalid_response`, same correlation rule as the SSE-framed path. Pinned by `test_json_response_with_mismatched_id_is_rejected`.

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 10]
```diff
@@ -0,0 +1,1049 @@
...
+                await self._notify("notifications/initialized", {})
+            except Exception:
+                await self._close_client()
+                raise
+            self._initialized = True
+
+    async def close(self) -> None:
+        """Terminate the MCP session and release the HTTP client."""
+        async with self._connect_lock:
+            await self._close_client()
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py:406]: Fixed in 27cd027555 — both `close()` implementations acquire `_request_lock` before `_connect_lock`, so teardown waits for the in-flight request. Bounded waiting is guaranteed because every request path is now under a total `asyncio.wait_for` deadline (finding #8), so close cannot block indefinitely.

File: apps/mcp-unified/src/mcp_unified/storage/models.py [hunk 11]
```diff
@@ -139,10 +139,11 @@ class ExternalServerDefinition(BaseModel):
 
     id: str
     name: str
-    transport: Literal["stdio", "websocket"]
+    transport: Literal["stdio", "websocket", "streamable_http", "sse"]
     command: list[str] = Field(default_factory=list)
     url: str | None = None
     cwd: str | None = None
+    headers: dict[str, str] = Field(default_factory=dict)
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/storage/models.py:146]: Fixed in 27cd027555 — `PatchExternalServerRequest` now admits `streamable_http`/`sse` transports and a `headers` field, so rotation and transport migration pass FastAPI validation and reach the manager's allowlist. Pinned by `test_patch_request_model_accepts_http_transports_and_headers`.

File: apps/mcp-unified/src/mcp_unified/federation/http_transport.py [hunk 12]
```diff
@@ -0,0 +1,1049 @@
...
+            params: dict[str, Any] = {
+                "name": tool_name,
+                "arguments": deepcopy(arguments or {}),
+            }
+            response = await self._request(
+                "tools/call",
+                params,
+                raise_on_error=False,
+                extra_headers=self._runtime_auth_headers(runtime_auth),
+            )
```
[the remark this answers is not in the document]
Comment (rmusser01) [reply apps/mcp-unified/src/mcp_unified/federation/http_transport.py:497]: Fixed in 27cd027555 — `_runtime_auth_headers` now enforces the same non-loopback plain-http refusal for brokered credential headers (`insecure_url`), and `call_tool` evaluates it before any connection attempt so nothing is transmitted first. Pinned by `test_brokered_credentials_over_plain_http_require_loopback` against a TEST-NET address.
